### PR TITLE
Thneed load/save

### PR DESCRIPTION
**Pull request by geohot**
Will be able to compile the dlc to thneed files.

This halves the size of the model in release. This also starts modeld faster.

### CODE REVIEW

**Comment by geohot** [comment]
Hmm, the model replay test is not expected to fail.

**Comment by geohot** [comment]
@adeebshihadeh Can you look into why tests are failing here? I expect the model to be exactly identical.

Could it be a race now that modeld starts up a lot faster?

**Comment by adeebshihadeh** [comment]
Test passes locally, but I can't reach jenkins remotely for some reason. Will check when in the office.

**Comment by geohot** [comment]
I'd been driving around with this, good to merge?

File: selfdrive/modeld/SConscript [hunk 1]
```diff
@@ -42,6 +48,14 @@ else:
 
 common = lenv.Object(common_src)
 
+# build thneed model
+if arch == "aarch64" or arch == "larch64":
+  compiler = lenv.Program('thneed/compile', ["thneed/compile.cc" ]+common, LIBS=libs)
```
Comment (adeebshihadeh) [inline selfdrive/modeld/SConscript:53]: this common shadows an imported common

File: selfdrive/modeld/runners/thneedmodel.h [hunk 2]
```diff
@@ -0,0 +1,33 @@
+#ifndef THNEEDMODEL_H
+#define THNEEDMODEL_H
+
+#include "runmodel.h"
+
+#define USE_CPU_RUNTIME 0
```
Comment (adeebshihadeh) [inline selfdrive/modeld/runners/thneedmodel.h:6]: these are also defined in snpemodel.h

File: selfdrive/modeld/runners/thneedmodel.h [hunk 3]
```diff
@@ -0,0 +1,33 @@
+#ifndef THNEEDMODEL_H
+#define THNEEDMODEL_H
```
Comment (adeebshihadeh) [inline selfdrive/modeld/runners/thneedmodel.h:2]: we've been replacing this style header guard with `#pragma once` across the codebase

**Comment by adeebshihadeh** [comment]
Still not sure why the outputs are different. Faster startup shouldn't matter since we only send each frame once we've got a model packet back. I generated new refs and ran the test a few times, but the output isn't deterministic.

File: selfdrive/modeld/thneed/compile.cc [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+#include "../runners/snpemodel.h"
+
+#define TEMPORAL_SIZE 512
+#define DESIRE_LEN 8
+#define TRAFFIC_CONVENTION_LEN 2
+
+// TODO: This should probably use SNPE directly.
+int main(int argc, char* argv[]) {
+  #define OUTPUT_SIZE 0x10000
+  float *output = (float*)calloc(OUTPUT_SIZE, sizeof(float));
```
Comment (adeebshihadeh) [inline selfdrive/modeld/thneed/compile.cc:12]: do we need to add TEMPORAL_SIZE here?

Comment (adeebshihadeh) [reply selfdrive/modeld/thneed/compile.cc:12]: oh, we're using a separate buffer but it's part of the output buffer in driving.cc

**Comment by geohot** [comment]
Can you check for nondeterminism again? There was a missing memset 0 for the recurrent state, maybe that was it?

**Comment by geohot** [comment]
@adeebshihadeh model replay test is passing, just "openpilot tests / longitudinal (push)" isn't

File: selfdrive/modeld/thneed/compile.cc (hunk 4 above)
Comment (geohot) [reply selfdrive/modeld/thneed/compile.cc:12]: nah, this is a junk buffer

**Comment by adeebshihadeh** [comment]
If you rebase the push and pull request tests should match.

**Comment by geohot** [comment]
Done

**Pull request merged**